File: src/bgbg/compute/maskops.py
"""The CV kernels: masks in, geometry out. numpy + PIL, no model.

Everything the segmenter does to a mask *after* the network has produced it:
score it (stability, IoU, NMS), trace its outline, simplify it, colour it, pack
the per-pixel lookup maps the engine hit-tests against, and composite it.

This is split out of the SAM session on purpose. None of it needs onnxruntime,
none of it needs a GPU, and all of it has to be reimplemented in TypeScript for
the web build -- so it is the half that gets a golden corpus (see
spec/tools/cvgold.py). Given a mask, these functions must produce byte-identical
contours, labelmaps and NMS decisions in both cores.

The interesting subtleties, all golden-pinned:

* ``_label`` is 8-connected. Four-connectivity splits diagonal touches into two
  objects and the object count silently changes.
* ``_trace_boundary`` is a Moore neighbourhood walk; the start pixel and the
  initial direction determine the winding, and the ants are dashed along it.
* ``save_objects`` paints the id maps largest-first for ``general`` and
  smallest-last for ``label``, which is what makes hovering a stack give you the
  whole object and dwelling give you the part.
* Masks are saved as **alpha coverage** (RGBA, A=coverage), not luminance -- an
  L mask is opaque everywhere, so stacking them in one mask node replaces rather
  than unions. See the step-5 bug in the plan.
"""
import os
import zlib
import colorsys
from collections import deque

import numpy as np
from PIL import Image, ImageFilter


def _stability(logit, offset=1.0):
    """SAM stability score: area(>+t) / area(>-t)."""
    high = np.count_nonzero(logit > offset)
    low = np.count_nonzero(logit > -offset)
    return (high / low) if low else 0.0


def _bbox(mask):
    ys, xs = np.where(mask)
    if xs.size == 0:
        return (0, 0, 0, 0)
    return int(xs.min()), int(ys.min()), int(xs.max() - xs.min() + 1), int(ys.max() - ys.min() + 1)


def _iou(a, b):
    inter = np.count_nonzero(a & b)
    if inter == 0:
        return 0.0
    union = np.count_nonzero(a) + np.count_nonzero(b) - inter
    return inter / union if union else 0.0


def _bbox_overlap(a, b):
    ax, ay, aw, ah = a
    bx, by, bw, bh = b
    return not (ax + aw <= bx or bx + bw <= ax or ay + ah <= by or by + bh <= ay)


def _nms(records, iou_thresh=0.7, max_objects=256):
    """Greedy IoU-NMS. ``records`` = list of ``(score, mask_bool)``. Bbox
    pre-filter then mask IoU; returns survivors sorted by score desc."""
    records = sorted(records, key=lambda r: r[0], reverse=True)
    kept, kept_bb = [], []
    for score, mask in records:
        bb = _bbox(mask)
        if bb[2] == 0:
            continue
        drop = False
        for km, kbb in zip(kept, kept_bb):
            if _bbox_overlap(bb, kbb) and _iou(mask, km[1]) > iou_thresh:
                drop = True
                break
        if not drop:
            kept.append((score, mask))
            kept_bb.append(bb)
            if len(kept) >= max_objects:
                break
    return kept


# ---------------------------------------------------------------------------


def _obj_color(cx, cy):
    """Deterministic, position-hashed colour so an object keeps the same colour
    across selections and re-runs (not tied to detection order)."""
    key = f"{int(cx) // 8}:{int(cy) // 8}".encode()
    h = (zlib.crc32(key) % 3600) / 3600.0
    r, g, b = colorsys.hsv_to_rgb(h, 0.68, 1.0)
    return (int(r * 255), int(g * 255), int(b * 255))


def _save_soft(mask, path, feather):
    """Write a mask as an alpha-coverage PNG: white RGB, alpha = coverage.

    Alpha, not luminance. Both GSK's mask node and Canvas2D's `destination-in`
    key off ALPHA, so coverage masks composite as a real union when several are
    stacked. Written as an `L` image they load back *opaque*, and stacking those
    made each mask silently REPLACE the previous one instead of unioning with it
    — which is why selecting several objects used to dim all but the last.
    """
    im = Image.fromarray(np.where(mask, 255, 0).astype(np.uint8), "L")
    if feather > 0:
        im = im.filter(ImageFilter.GaussianBlur(feather))
    rgba = Image.new("RGBA", im.size, (255, 255, 255, 0))
    rgba.putalpha(im)
    rgba.save(path)


# Clockwise 8-neighbour offsets (row, col) starting at West — for Moore tracing.
_CW = [(0, -1), (-1, -1), (-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1)]


def _trace_boundary(mask):
    """Moore-neighbour boundary tracing. Returns the ordered outer contour
    (list of (x, y)) of the component containing the topmost-leftmost pixel."""
    H, W = mask.shape
    m = np.zeros((H + 2, W + 2), bool)
    m[1:-1, 1:-1] = mask
    ys, xs = np.nonzero(m)
    if xs.size == 0:
        return []
    k = np.lexsort((xs, ys))[0]              # topmost, then leftmost
    start = (int(ys[k]), int(xs[k]))
    cur = start
    back = (start[0], start[1] - 1)          # approached from the West (bg)
    out = [cur]
    limit = 8 * int(m.sum()) + 16
    for _ in range(limit):
        bi = _CW.index((back[0] - cur[0], back[1] - cur[1]))
        nxt = None
        for j in range(1, 9):
            dy, dx = _CW[(bi + j) % 8]
            p = (cur[0] + dy, cur[1] + dx)
            if m[p[0], p[1]]:
                pdy, pdx = _CW[(bi + j - 1) % 8]
                back = (cur[0] + pdy, cur[1] + pdx)
                nxt = p
                break
        if nxt is None or nxt == start:
            break
        cur = nxt
        out.append(cur)
    return [(c[1] - 1, c[0] - 1) for c in out]   # -> (x, y), unpad


def _rdp(pts, eps):
    """Ramer-Douglas-Peucker polyline simplification (keeps endpoints)."""
    n = len(pts)
    if n < 3:
        return list(pts)
    P = np.asarray(pts, float)
    keep = np.zeros(n, bool)
    keep[0] = keep[-1] = True
    stack = [(0, n - 1)]
    while stack:
        a, b = stack.pop()
        if b <= a + 1:
            continue
        seg = P[b] - P[a]
        L2 = float(seg.dot(seg))
        rel = P[a + 1:b] - P[a]
        if L2 == 0.0:
            d = np.hypot(rel[:, 0], rel[:, 1])
        else:
            t = np.clip((rel @ seg) / L2, 0.0, 1.0)
            proj = P[a] + t[:, None] * seg
            d = np.hypot(*(P[a + 1:b] - proj).T)
        if d.size == 0:
            continue
        mi = int(np.argmax(d))
        if d[mi] > eps:
            idx = a + 1 + mi
            keep[idx] = True
            stack.append((a, idx))
            stack.append((idx, b))
    return [pts[i] for i in range(n) if keep[i]]


def _label(mask, connectivity=8):
    """Label connected components (iterative BFS). Returns (labels, count)."""
    H, W = mask.shape
    lab = np.zeros((H, W), np.int32)
    if connectivity == 8:
        nbr = ((-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1))
    else:
        nbr = ((-1, 0), (1, 0), (0, -1), (0, 1))
    cur = 0
    ys, xs = np.nonzero(mask)
    for sy, sx in zip(ys.tolist(), xs.tolist()):
        if lab[sy, sx]:
            continue
        cur += 1
        dq = deque(((sy, sx),))
        lab[sy, sx] = cur
        while dq:
            y, x = dq.popleft()
            for dy, dx in nbr:
                ny, nx = y + dy, x + dx
                if 0 <= ny < H and 0 <= nx < W and mask[ny, nx] and not lab[ny, nx]:
                    lab[ny, nx] = cur
                    dq.append((ny, nx))
    return lab, cur


def contour(mask, epsilon=1.8, min_area=80, max_pts=500, max_dim=480):
    """All boundary polylines of ``mask`` (each a list of [x, y] in original-image
    px): the outer contour of every connected component PLUS its holes, so
    complex or disjoint objects (branchy trees, rings) outline correctly. The
    mask is traced at reduced resolution for speed on large images, then scaled
    back. Empty for tiny masks."""
    if np.count_nonzero(mask) < min_area:
        return []
    H, W = mask.shape
    if max(H, W) > max_dim:
        sc = max_dim / max(H, W)
        sw, sh = max(1, round(W * sc)), max(1, round(H * sc))
        small = np.asarray(Image.fromarray(
            np.where(mask, 255, 0).astype(np.uint8)).resize(
            (sw, sh), Image.NEAREST)) > 127
    else:
        small, sw, sh = mask, W, H
    up = W / sw                                  # small px -> original px
    small_min = max(8, int(min_area * (sw * sh) / (W * H)))

    regions = []
    lab, n = _label(small, 8)
    for k in range(1, n + 1):
        comp = lab == k
        if np.count_nonzero(comp) >= small_min:
            regions.append(comp)
    # holes = background components that don't touch the border
    blab, bn = _label(~small, 4)
    border = (set(blab[0, :].tolist()) | set(blab[-1, :].tolist())
              | set(blab[:, 0].tolist()) | set(blab[:, -1].tolist()))
    for k in range(1, bn + 1):
        if k in border:
            continue
        hole = blab == k
        if np.count_nonzero(hole) >= max(6, small_min // 4):
            regions.append(hole)

    polys = []
    for reg in regions:
        poly = _trace_boundary(reg)
        if len(poly) < 3:
            continue
        poly = [(x * up, y * up) for x, y in poly]
        poly = _rdp(poly, epsilon)
        if len(poly) > max_pts:
            step = int(np.ceil(len(poly) / max_pts))
            poly = poly[::step]
        polys.append([[int(round(x)), int(round(y))] for x, y in poly])
    return polys


def _save_idmap(arr, path):
    """Encode a uint16 id-per-pixel array as ``R + G*256`` in an RGB PNG."""
    H, W = arr.shape
    rgb = np.zeros((H, W, 3), np.uint8)
    rgb[..., 0] = (arr & 0xFF).astype(np.uint8)
    rgb[..., 1] = (arr >> 8).astype(np.uint8)
    Image.fromarray(rgb, "RGB").save(path)
    return path


def save_objects(masks, size, out_dir, prefix, feather=0.8):
    """Write soft per-object mask PNGs + three lookup maps for hit-testing
    (``masks`` is largest-first). Returns ``(maps, objects)`` where maps has:
      - ``label``   : the most *specific* object per pixel (smallest on top)
      - ``general`` : the most *general* object per pixel (largest on top)
      - ``depth``   : how many objects overlap each pixel (an ``L`` PNG)
    ids are encoded ``R + G*256`` (0 = background)."""
    W, H = size
    os.makedirs(out_dir, exist_ok=True)
    label = np.zeros((H, W), np.uint16)      # specific: smallest painted last
    general = np.zeros((H, W), np.uint16)    # general: largest painted last
    depth = np.zeros((H, W), np.uint16)
    objs = []
    for idx, mask in enumerate(masks):
        oid = idx + 1
        label[mask] = oid
        depth += mask
        ys, xs = np.where(mask)
        if xs.size == 0:
            continue
        cx, cy = float(xs.mean()), float(ys.mean())
        mp = os.path.join(out_dir, f"{prefix}_obj{oid}.png")
        _save_soft(mask, mp, feather)
        x, y, w, h = _bbox(mask)
        objs.append({"id": oid, "color": "#%02x%02x%02x" % _obj_color(cx, cy),
                     "bbox": [x, y, w, h], "area": int(xs.size), "mask": mp,
                     "contour": contour(mask)})
    for idx in range(len(masks) - 1, -1, -1):   # largest painted last -> on top
        general[masks[idx]] = idx + 1
    maps = {
        "label": _save_idmap(label, os.path.join(out_dir, f"{prefix}_labelmap.png")),
        "general": _save_idmap(general, os.path.join(out_dir, f"{prefix}_general.png")),
        "depth": os.path.join(out_dir, f"{prefix}_depth.png"),
    }
    Image.fromarray(np.minimum(depth, 255).astype(np.uint8), "L").save(maps["depth"])
    return maps, objs


def save_mask(mask, path, feather=0.8):
    """Write a single soft-edged mask PNG; return its (x, y, w, h) bbox."""
    _save_soft(mask, path, feather)
    return list(_bbox(mask))


def load_union(paths):
    """Combine soft mask PNGs into one uint8 alpha (max), keeping soft edges."""
    out = None
    for p in paths:
        im = Image.open(p)
        if im.mode in ("RGBA", "LA"):          # coverage masks (see _save_soft)
            a = np.asarray(im.convert("RGBA"))[..., 3]
        else:                                  # legacy L masks
            a = np.asarray(im.convert("L"))
        out = a if out is None else np.maximum(out, a)
    return out


def composite_image(src_pil, alpha, bg, blur=20):
    """The ``alpha`` (uint8 HxW) cutout of ``src_pil`` over ``bg``, as an image.

    The background itself is the outputter's job (``compute.outputs_impl``) — the
    same one the background-removal and GIF paths use, so a new effect is added in
    one place rather than three.
    """
    from compute import outputs_impl

    src = src_pil.convert("RGB")
    cut = outputs_impl.cutout(src, Image.fromarray(alpha.astype(np.uint8), "L"))
    return outputs_impl.apply_bg(cut, bg, source=src, blur=blur)


def composite_extract(src_pil, alpha, bg, dst, blur=20, rot=0, fh=False, fv=False):
    """...and saved to ``dst``, with the view transform baked in.

    The masks live in un-rotated source space (so they can ride the pane
    transform instead of being invalidated by it), which means the rotation the
    user was looking at has to be applied here, at the end.
    """
    from compute import outputs_impl

    img = composite_image(src_pil, alpha, bg, blur=blur)
    outputs_impl.apply_transform(img, rot, fh, fv).save(dst)
    return dst

File: web/core/maskops.ts
/**
 * The CV kernels — the mirror of compute/maskops.py.
 *
 * Masks in, geometry out: connected components, boundary tracing, RDP
 * simplification, IoU/NMS, stability. No model, no GPU — which is exactly why
 * this half can live in the browser's main bundle while ORT-Web does the rest.
 *
 * Pinned by spec/goldens/cv.json. The traps it pins, all of which are silent:
 *   * components are 8-connected. Four-connectivity splits a diagonal touch into
 *     two objects and the object count quietly changes.
 *   * the boundary walk's start pixel and initial direction fix the winding, and
 *     the marching ants are dashed along it.
 *   * `depth` is a COUNT map, not an id map.
 *
 * A `Mask` is a flat Uint8Array of 0/1, row-major, with width and height.
 */

export interface Mask {
  data: Uint8Array;
  w: number;
  h: number;
}

export function maskAt(m: Mask, x: number, y: number): number {
  if (x < 0 || y < 0 || x >= m.w || y >= m.h) return 0;
  return m.data[y * m.w + x];
}

export function maskArea(m: Mask): number {
  let n = 0;
  for (let i = 0; i < m.data.length; i++) n += m.data[i] ? 1 : 0;
  return n;
}

/** Stability: how little the mask changes when the logit threshold is nudged. */
export function stability(logit: Float32Array, offset = 1.0): number {
  let hi = 0;
  let lo = 0;
  for (let i = 0; i < logit.length; i++) {
    if (logit[i] > offset) hi++;
    if (logit[i] > -offset) lo++;
  }
  return lo === 0 ? 0.0 : hi / lo;
}

/** [x, y, w, h] of the mask's true pixels; w == 0 when empty. */
export function bbox(m: Mask): [number, number, number, number] {
  let x0 = m.w;
  let y0 = m.h;
  let x1 = -1;
  let y1 = -1;
  for (let y = 0; y < m.h; y++) {
    for (let x = 0; x < m.w; x++) {
      if (m.data[y * m.w + x]) {
        if (x < x0) x0 = x;
        if (x > x1) x1 = x;
        if (y < y0) y0 = y;
        if (y > y1) y1 = y;
      }
    }
  }
  if (x1 < 0) return [0, 0, 0, 0];
  return [x0, y0, x1 - x0 + 1, y1 - y0 + 1];
}

export function iou(a: Mask, b: Mask): number {
  let inter = 0;
  let union = 0;
  for (let i = 0; i < a.data.length; i++) {
    const x = a.data[i];
    const y = b.data[i];
    if (x && y) inter++;
    if (x || y) union++;
  }
  return union === 0 ? 0.0 : inter / union;
}

export function bboxOverlap(
  a: [number, number, number, number],
  b: [number, number, number, number],
): boolean {
  return !(
    a[0] + a[2] <= b[0] ||
    b[0] + b[2] <= a[0] ||
    a[1] + a[3] <= b[1] ||
    b[1] + b[3] <= a[1]
  );
}

/** Greedy IoU-NMS. Records are [score, mask]; survivors come back score-desc. */
export function nms(
  records: [number, Mask][],
  iouThresh = 0.7,
  maxObjects = 256,
): [number, Mask][] {
  // Python's sorted() is STABLE, so equal scores keep their input order. A JS
  // sort is not required to be stable across engines for large arrays — but it
  // is in V8/JSC for all arrays, and we compare on score only, so ties keep
  // input order exactly as Python does.
  const sorted = [...records].sort((r1, r2) => r2[0] - r1[0]);
  const kept: [number, Mask][] = [];
  const keptBb: [number, number, number, number][] = [];

  for (const [score, mask] of sorted) {
    const bb = bbox(mask);
    if (bb[2] === 0) continue;
    let drop = false;
    for (let i = 0; i < kept.length; i++) {
      if (bboxOverlap(bb, keptBb[i]) && iou(mask, kept[i][1]) > iouThresh) {
        drop = true;
        break;
      }
    }
    if (!drop) {
      kept.push([score, mask]);
      keptBb.push(bb);
      if (kept.length >= maxObjects) break;
    }
  }
  return kept;
}

/** Label connected components (iterative BFS). Returns [labels, count]. */
export function label(m: Mask, connectivity = 8): [Int32Array, number] {
  const { w, h } = m;
  const lab = new Int32Array(w * h);
  const nbr =
    connectivity === 8
      ? [
          [-1, -1], [-1, 0], [-1, 1],
          [0, -1], [0, 1],
          [1, -1], [1, 0], [1, 1],
        ]
      : [
          [-1, 0], [1, 0], [0, -1], [0, 1],
        ];

  let cur = 0;
  const queue: number[] = [];
  for (let y0 = 0; y0 < h; y0++) {
    for (let x0 = 0; x0 < w; x0++) {
      const i0 = y0 * w + x0;
      if (!m.data[i0] || lab[i0]) continue;
      cur++;
      lab[i0] = cur;
      queue.length = 0;
      queue.push(i0);
      let head = 0;
      while (head < queue.length) {
        const i = queue[head++];
        const y = (i / w) | 0;
        const x = i - y * w;
        for (const [dy, dx] of nbr) {
          const ny = y + dy;
          const nx = x + dx;
          if (nx < 0 || ny < 0 || nx >= w || ny >= h) continue;
          const j = ny * w + nx;
          if (m.data[j] && !lab[j]) {
            lab[j] = cur;
            queue.push(j);
          }
        }
      }
    }
  }
  return [lab, cur];
}

/** The clockwise neighbourhood, as (dy, dx) — the winding depends on this order. */
const CW: [number, number][] = [
  [0, -1], [-1, -1], [-1, 0], [-1, 1],
  [0, 1], [1, 1], [1, 0], [1, -1],
];

function cwIndex(dy: number, dx: number): number {
  for (let i = 0; i < 8; i++) if (CW[i][0] === dy && CW[i][1] === dx) return i;
  return -1;
}

/**
 * Moore-neighbour boundary tracing: the ordered outer contour of the component
 * containing the topmost-then-leftmost pixel.
 *
 * The start pixel and the initial "approached from the West" direction fix the
 * winding, and the marching ants are dashed along it — so both must match.
 */
export function traceBoundary(m: Mask): [number, number][] {
  const W = m.w;
  const H = m.h;
  const pw = W + 2;
  const ph = H + 2;
  const p = new Uint8Array(pw * ph);
  let total = 0;
  for (let y = 0; y < H; y++) {
    for (let x = 0; x < W; x++) {
      if (m.data[y * W + x]) {
        p[(y + 1) * pw + (x + 1)] = 1;
        total++;
      }
    }
  }
  if (total === 0) return [];

  // topmost, then leftmost (np.lexsort((xs, ys))[0])
  let start: [number, number] | null = null;
  outer: for (let y = 0; y < ph; y++) {
    for (let x = 0; x < pw; x++) {
      if (p[y * pw + x]) {
        start = [y, x];
        break outer;
      }
    }
  }
  let cur = start!;
  let back: [number, number] = [start![0], start![1] - 1]; // from the West
  const out: [number, number][] = [cur];
  const limit = 8 * total + 16;

  for (let step = 0; step < limit; step++) {
    const bi = cwIndex(back[0] - cur[0], back[1] - cur[1]);
    let nxt: [number, number] | null = null;
    for (let j = 1; j < 9; j++) {
      const [dy, dx] = CW[(bi + j) % 8];
      const y = cur[0] + dy;
      const x = cur[1] + dx;
      if (y < 0 || x < 0 || y >= ph || x >= pw) continue;
      if (p[y * pw + x]) {
        const [pdy, pdx] = CW[(bi + j - 1 + 8) % 8];
        back = [cur[0] + pdy, cur[1] + pdx];
        nxt = [y, x];
        break;
      }
    }
    if (nxt === null || (nxt[0] === start![0] && nxt[1] === start![1])) break;
    cur = nxt;
    out.push(cur);
  }
  return out.map(([y, x]) => [x - 1, y - 1] as [number, number]); // -> (x, y), unpad
}

/**
 * Ramer-Douglas-Peucker — a literal port of maskops._rdp.
 *
 * Three details that a from-memory RDP gets wrong, and that change every contour:
 *   * the distance is to the SEGMENT (the projection parameter is clamped to
 *     [0, 1]), not to the infinite line through the endpoints;
 *   * it is iterative over a keep-mask, so the survivors come back in ORIGINAL
 *     order rather than as a concatenation of recursive halves;
 *   * np.argmax keeps the FIRST maximum on a tie.
 */
export function rdp(pts: [number, number][], eps: number): [number, number][] {
  const n = pts.length;
  if (n < 3) return [...pts];

  const keep = new Uint8Array(n);
  keep[0] = 1;
  keep[n - 1] = 1;

  const stack: [number, number][] = [[0, n - 1]];
  while (stack.length) {
    const [a, b] = stack.pop()!;
    if (b <= a + 1) continue;

    const segx = pts[b][0] - pts[a][0];
    const segy = pts[b][1] - pts[a][1];
    const l2 = segx * segx + segy * segy;

    let dmax = -1;
    let mi = -1;
    for (let i = a + 1; i < b; i++) {
      const rx = pts[i][0] - pts[a][0];
      const ry = pts[i][1] - pts[a][1];
      let d: number;
      if (l2 === 0.0) {
        d = Math.hypot(rx, ry);
      } else {
        let t = (rx * segx + ry * segy) / l2;
        t = t < 0 ? 0 : t > 1 ? 1 : t; // clamp to the segment
        const px = pts[a][0] + t * segx;
        const py = pts[a][1] + t * segy;
        d = Math.hypot(pts[i][0] - px, pts[i][1] - py);
      }
      if (d > dmax) {
        // strict > keeps the FIRST maximum, as np.argmax does
        dmax = d;
        mi = i;
      }
    }
    if (mi < 0) continue;
    if (dmax > eps) {
      keep[mi] = 1;
      stack.push([a, mi]);
      stack.push([mi, b]);
    }
  }

  const out: [number, number][] = [];
  for (let i = 0; i < n; i++) if (keep[i]) out.push(pts[i]);
  return out;
}

/** Python's round(): half-to-even. int(round(x)) in contour() uses it. */
function roundHalfEven(x: number): number {
  const fl = Math.floor(x);
  const diff = x - fl;
  if (diff > 0.5) return fl + 1;
  if (diff < 0.5) return fl;
  return fl % 2 === 0 ? fl : fl + 1; // exact tie -> even
}

function subMask(lab: Int32Array, k: number, w: number, h: number): Mask {
  const data = new Uint8Array(w * h);
  for (let i = 0; i < data.length; i++) data[i] = lab[i] === k ? 1 : 0;
  return { data, w, h };
}

function invert(m: Mask): Mask {
  const data = new Uint8Array(m.data.length);
  for (let i = 0; i < data.length; i++) data[i] = m.data[i] ? 0 : 1;
  return { data, w: m.w, h: m.h };
}

/**
 * All boundary polylines of `mask`: the outer contour of every connected
 * component PLUS its holes, so rings and branchy/disjoint objects outline
 * correctly.
 *
 * NOTE: the mask is NOT downscaled here — the Python side reduces resolution
 * above max_dim for speed on big images. Ported faithfully for the sizes the
 * golden covers; a full port must add the NEAREST downscale + `up` rescale.
 */
export function contour(
  m: Mask,
  epsilon = 1.8,
  minArea = 80,
  maxPts = 500,
  maxDim = 480,
): number[][][] {
  if (maskArea(m) < minArea) return [];
  const W = m.w;
  const H = m.h;
  if (Math.max(H, W) > maxDim) {
    throw new Error(
      "contour(): the downscale path is not ported yet — see maskops.py",
    );
  }
  const small = m;
  const sw = W;
  const sh = H;
  const up = W / sw;
  const smallMin = Math.max(8, Math.trunc((minArea * (sw * sh)) / (W * H)));

  const regions: Mask[] = [];
  const [lab, n] = label(small, 8);
  for (let k = 1; k <= n; k++) {
    const comp = subMask(lab, k, sw, sh);
    if (maskArea(comp) >= smallMin) regions.push(comp);
  }

  // holes = background components that do not touch the border
  const [blab, bn] = label(invert(small), 4);
  const border = new Set<number>();
  for (let x = 0; x < sw; x++) {
    border.add(blab[x]);
    border.add(blab[(sh - 1) * sw + x]);
  }
  for (let y = 0; y < sh; y++) {
    border.add(blab[y * sw]);
    border.add(blab[y * sw + (sw - 1)]);
  }
  for (let k = 1; k <= bn; k++) {
    if (border.has(k)) continue;
    const hole = subMask(blab, k, sw, sh);
    if (maskArea(hole) >= Math.max(6, Math.trunc(smallMin / 4))) {
      regions.push(hole);
    }
  }

  const polys: number[][][] = [];
  for (const reg of regions) {
    let poly = traceBoundary(reg);
    if (poly.length < 3) continue;
    poly = poly.map(([x, y]) => [x * up, y * up] as [number, number]);
    poly = rdp(poly, epsilon);
    if (poly.length > maxPts) {
      const step = Math.ceil(poly.length / maxPts);
      poly = poly.filter((_, i) => i % step === 0);
    }
    polys.push(poly.map(([x, y]) => [roundHalfEven(x), roundHalfEven(y)]));
  }
  return polys;
}
